activerecord/lib/active_record/attributes.rb

```diff
@@ -208,21 +208,30 @@ module ClassMethods
       # tracking is performed. The methods +changed?+ and +changed_in_place?+
       # will be called from ActiveModel::Dirty. See the documentation for those
       # methods in ActiveModel::Type::Value for more details.
-      def attribute(name, cast_type = nil, **options, &decorator)
+      def attribute(name, cast_type = nil, default: NO_DEFAULT_PROVIDED, **options, &block)
         name = name.to_s
         reload_schema_from_cache
 
-        prev_cast_type, prev_options, prev_decorator = attributes_to_define_after_schema_loads[name]
+        case cast_type
+        when Symbol
+          type = cast_type
+          cast_type = -> _ { Type.lookup(type, **options, adapter: Type.adapter_name_from(self)) }
+        when nil
+          if (prev_cast_type, prev_default = attributes_to_define_after_schema_loads[name])
+            default = prev_default if default == NO_DEFAULT_PROVIDED
 
-        unless cast_type && prev_cast_type
-          cast_type ||= prev_cast_type
-          options = prev_options || options if options.empty?
-          decorator ||= prev_decorator
+            cast_type = if block_given?
+              -> subtype { yield Proc === prev_cast_type ? prev_cast_type[subtype] : prev_cast_type }
+            else
+              prev_cast_type
+            end
+          else
+            cast_type = block || -> subtype { subtype }
+          end
         end
 
-        self.attributes_to_define_after_schema_loads = attributes_to_define_after_schema_loads.merge(
-          name => [cast_type, options, decorator]
-        )
+        self.attributes_to_define_after_schema_loads =
+          attributes_to_define_after_schema_loads.merge(name => [cast_type, default])
```

Comment:
Can we assume that `attributes_to_define_after_schema_loads[name][0]` will now always be either a `Proc` or `nil`?  If so, what about:

```ruby
prev_type_proc, prev_default = attributes_to_define_after_schema_loads[name]

type_proc =
  if cast_type
    -> _ { Type.lookup(cast_type, **options, adapter: Type.adapter_name_from(self)) }
  else
    [prev_type_proc, block].compact.reduce { |f, g| -> subtype { g[f[subtype]] } }
  end

default = prev_default if cast_type.nil? && default == NO_DEFAULT_PROVIDED

self.attributes_to_define_after_schema_loads =
  attributes_to_define_after_schema_loads.merge(name => [type_proc, default])
```

The `else` condition could be written more neatly as `[prev_type_proc, block].compact.reduce(&:>>)` when we drop Ruby 2.5 support, or if we add a `Proc#>>` polyfill.

Reply:
`cast_type` will always be `Proc` or a type object, isn't `nil`.

Reply:
i.e. `attribute :attr_name, EncryptedType.new` should work as it is.

Reply:
Ah, right!  I forgot that case.  So then perhaps:

```ruby
prev_type_proc, prev_default = attributes_to_define_after_schema_loads[name]

type_proc =
  case cast_type
  when nil
    [prev_type_proc, block].compact.reduce { |f, g| -> subtype { g[f[subtype]] } }
  when Symbol
    -> _ { Type.lookup(cast_type, **options, adapter: Type.adapter_name_from(self)) }
  else
    -> _ { cast_type }
  end

default = prev_default if cast_type.nil? && default == NO_DEFAULT_PROVIDED

self.attributes_to_define_after_schema_loads =
  attributes_to_define_after_schema_loads.merge(name => [type_proc, default])
```

Reply:
`default = prev_default if cast_type.nil? && default == NO_DEFAULT_PROVIDED` in the above will override database default to `nil` accidentally, and `[prev_type_proc, block].compact.reduce { |f, g| -> subtype { g[f[subtype]] } }` can be `nil`.

Personally I'm not interested creating extra block especially when `cast_type` is a symbol or a type object which are the most case, just to use `Proc#>>` in a future.

(Wrapping a block for a symbol is to avoid `bin/test` testing issue. I'll fix that in a later PR.)

Reply:
> and `[prev_type_proc, block].compact.reduce { |f, g| -> subtype { g[f[subtype]] } }` can be `nil`.

Yes.  It would require changing `load_schema!` from

```ruby
cast_type = cast_type[type_for_attribute(name)] if Proc === cast_type
```

to something like

```ruby
cast_type = type_for_attribute(name)
cast_type = type_proc[cast_type] if type_proc
```

> `default = prev_default if cast_type.nil? && default == NO_DEFAULT_PROVIDED` in the above will override database default to `nil` accidentally,

That could be addressed by adding a default value of `[nil, NO_DEFAULT_PROVIDED]` to the `attributes_to_define_after_schema_loads` Hash.


However, it seems like you have plans for this code, so let's do the way that you prefer! :+1: